cmd/secrets-store-csi-driver/main.go

```diff
@@ -62,6 +61,12 @@ var (
 	enableProfile          = flag.Bool("enable-pprof", false, "enable pprof profiling")
 	profilePort            = flag.Int("pprof-port", 6065, "port for pprof profiling")
 
+	// enable filtered watch for secrets. The filtering is done on the csi driver label: secrets-store.csi.k8s.io/managed=true
+	// this label is set for all Kubernetes secrets created by the CSI driver. For Kubernetes secrets used to provide credentials
+	// for use with the CSI driver, set the label by running: kubectl label secret secrets-store-creds secrets-store.csi.k8s.io/managed=true
+	// This feature flag will be enabled by default after n+2 releases giving time for users to label all their existing credential secrets.
+	filteredWatchSecret = flag.Bool("filtered-watch-secret", false, "enable filtered watch for secrets with label secrets-store.csi.k8s.io/managed=true")
```

Comment:
To make sure I understand this right, all sync'd secrets should _already_ have the `secrets-store-creds secrets-store.csi.k8s.io/managed=true` label right? This flag is just for controlling updating `nodePublishSecretRef` secrets to also have the label?

It feels wrong to me that the `secrets-store-creds secrets-store.csi.k8s.io/managed=true` label would need to be added to the `nodePublishSecretRef` secrets, since the driver isn't really managing those secrets.

Do we need to watch these secrets?

Reply:
The cache is initialized with the label selector if the flag is enabled. This would mean only secrets with `secrets-store.csi.k8s.io/managed=true` label will be cached. If we try to get the `nodePublishSecretRef` from the cache, it'll not exist. This is required for rotation reconciler. Otherwise we'll be doing a `GET` call periodically to the API server for the `nodePublishSecretRef`. This will result in large number of calls if all the pods are using `nodePublishSecretRef`.

Reply:
I think perhaps the word `managed` in `secrets-store-creds secrets-store.csi.k8s.io/managed=true` feels weird for `nodePublishSecretRef` secrets as they are not managed by the driver, but rather consumed. We could update the label for `nodePublishSecretRef` secret as `secrets-store-creds secrets-store.csi.k8s.io/used=true`.

Since adding labels to `nodePublishSecretRef` secret requires an update to existing resources, if they fail to add the label, then we would not be able to find the secret and cache it.

Reply:
> I think perhaps the word managed in secrets-store-creds secrets-store.csi.k8s.io/managed=true feels weird for nodePublishSecretRef secrets as they are not managed by the driver, but rather consumed. We could update the label for nodePublishSecretRef secret as secrets-store-creds secrets-store.csi.k8s.io/used=true.

Kubernetes doesn't allow OR for the label selector which means we can do filtered watch only a single label while initializing the cache.

> Since adding labels to nodePublishSecretRef secret requires an update to existing resources, if they fail to add the label, then we would not be able to find the secret and cache it.

Yeah, this is a breaking change. That's why it's not enabled by default. We'll add a note that the feature will be enabled by default in a future release and necessary action (labelling existing secrets) need to be done.

Reply:
The `NodePublishSecretRef` is really only required during rotation. We should be able to use a separate cache with filter on `secrets-store.csi.k8s.io/used=true` label. I've added a [commit](https://github.com/kubernetes-sigs/secrets-store-csi-driver/pull/458/commits/5688c9cc2a651fdd88dd9b0a5a972ce9a643bebd) that does that. I'll run tests to see what the impact is with the change.

Gist of the changes:
1. We enable filtered watch for secrets created by the CSI driver by default
2. User can enable filtered watch for node publish secret ref after labelling existing secrets and then enable `--filtered-watch-secret`

Reply:
Great! I think this would resolve my concerns if the test results look promising.

Reply:
Yeah, the test results look good. I've updated the proposal with the latest results.